src/new_particle_filter.py

```diff
@@ -0,0 +1,161 @@
+                particles[:, i] %= 2 * np.pi
+
+        self.particles = particles
+        self.weights = np.asmatrix(np.ones((N, 1)))
+        self.weights /= sum(self.weights)
+
+    def correct(self, *args):
+        np.multiply(self.weights, self.MeasurementLikelihoodFcn(
+            self.particles, *args))
+
+        self.weights += 1.e-100
+        self.weights /= sum(self.weights)
+
+        N = np.shape(self.weights)[0]
+        if self.neff(self.weights) < N/2:
+            indexes = systematic_resample(self.weights)
+            self.particles, self.weights = self.resample_from_index(
+                self.particles, self.weights, indexes)
+
+        return self.estimate(self.particles, self.weights)
+
+    # u should be a column matrix
+    def predict(self, dt, u):
+        if self.StateTransitionFcn is None:
+            raise Exception(
+                "The particle filter has not been \
+                 given a state transition function")
+
+        self.particles = self.StateTransitionFcn(self.particles, dt, u)
+
+    def estimate(self, particles, weights):
+        mean = np.average(
+            particles, weights=np.asarray(weights.T)[0], axis=0).T
+        var = np.average(np.power((particles - mean.T), 2),
+                         weights=np.asarray(weights.T)[0], axis=0)
+        return mean, var
+
+    def resample_from_index(self, particles, weights, indexes):
+        particles[:] = particles[indexes]
+        weights[:] = weights[indexes]
+        weights.fill(1.0 / len(weights))
+
+        return particles, weights
+
+    def multinomial_resample(self, particles, weights):
+        N = len(particles)
+        cumulative_sum = np.cumsum(weights)
+        cumulative_sum[-1] = 1.
+        indexes = np.searchsorted(cumulative_sum, uniform(0, 1, N))
+
+        particles[:] = particles[indexes]
+        weights.fill(1.0 / N)
+        return particles, weights
+
+    def neff(self, weights):
+        return 1. / np.sum(np.square(weights))
+
+
+if __name__ == "__main__":
+    pf = ParticleFilter()
+
+    Ncars = 3
+    Nparticles = 1
+    total_time = 5.
+    dt = 0.05
+
+    initial_positions = np.matrix(
+        [[0, 0, np.pi / 2], [1, 1, -np.pi / 2], [-1, -2, 0]])
+    initial_transforms = np.asmatrix(np.zeros(((Ncars - 1), 3)))
+    circ_var = [0, 0, 1, 0, 0, 1]
+    limits = np.matrix([0.5, 0.5, np.pi / 2, 0.5, 0.5, np.pi / 2]).T
+    bounds = np.asmatrix(np.zeros((3 * (Ncars - 1), 2)))
+
+    for i in range(Ncars - 1):
+        init_xy = pf.rot(-initial_positions[0, 2]) * \
+            initial_positions[i + 1, :2].T
+        init_theta = initial_positions[i + 1, 2] - initial_positions[0, 2]
+        initial_transforms[i, :] = np.append(init_xy.T, [[init_theta]], axis=1)
+
+        fi = 3 * i
+        bounds[fi:fi + 3, 0] = initial_transforms[i, :].T - limits[fi:fi + 3]
+        bounds[fi:fi + 3, 1] = initial_transforms[i, :].T + limits[fi:fi + 3]
+
+    noise_u = np.matrix([0.1, 0.05, 0.1, 0.05, 0.1, 0.05])
```

Comment:
is this the input noise variance? If yes, maybe rename to `input_noise_var`.  Maybe worth declaring at beginning of the function?